Implement basics of link preloading

TimvdLippe opened the pull request:
These changes allow a minimal set of checks for font-src
CSP checks to pass.

Part of #4577
Part of #35035

TimvdLippe committed:
Add baseline for preload WPT tests

Part of #35035

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
I split this PR up in 2 commits, so that you can see which WPT tests would be fixed with the actual implementation. The reason I didn't submit 2 PR's is that as soon as you implement the basics, a lot of WPT tests start to FAIL instead of TIMEOUT. Therefore, we end up changing the expectations twice resulting in a lot of noise.

TimvdLippe commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -494,6 +507,128 @@ impl HTMLLinkElement {
+            let this = Trusted::new(self);
+            self.owner_global()
+                .task_manager()
+                .performance_timeline_task_source()
```
I don't understand why this would be necessary and in fact, it's not even working yet. The test I use is `tests/wpt/tests/preload/preload-referrer-policy.html`. I can confirm that it preloads fine and it loads the script. Also, the performance timing entries are there if you add the following:

```diff
--- a/tests/wpt/tests/preload/preload-referrer-policy.html
+++ b/tests/wpt/tests/preload/preload-referrer-policy.html
@@ -1,7 +1,7 @@
 <!DOCTYPE html>
 <meta charset=utf-8>
 <title>The preload's referrerpolicy attribute should be respected</title>
-<meta name="timeout" content="long">
+<!-- <meta name="timeout" content="long"> -->
 <script src="resources/dummy.js?link-header-preload2"></script>
 <script src="/common/get-host-info.sub.js"></script>
 <script src="/common/utils.js"></script>
@@ -52,6 +52,8 @@ const loaders = {
         const loaded = new Promise(resolve => script.addEventListener('load', resolve));
         document.body.appendChild(script);
         await loaded;
+        console.log(performance.getEntries());
+        console.log(script.src);
         return {unsafe: location.href, actualReferrer: window.referrers[id], entries: performance.getEntriesByName(script.src).length}
     },
 };
```

However, the performance timings of run 1 are there, when executing run 2. That means that when the `load` event fires on the link, then the performance timing entries haven't been committed yet. Hence the `performance.getEntriesByName(script.src).length` results in 0. If you were to run that code in the second run with the `src` of the first run, it would show up data as expected.

I though the `performance_timeline_task_source` was the trick here to queue a task to ensure the rest has already been commited, but that's not enough.

TimvdLippe commented on components/shared/net/mime_classifier.rs:
```diff
@@ -3,7 +3,8 @@
+
+use crate::LoadContext;
 
 pub struct MimeClassifier {
```
I moved this into `net_traits` since `script` does not depend on `net`. Since it is a relatively clean class with barely any dependencies, I suppose that's fine?

TimvdLippe commented:
Let's see what WPT says: https://github.com/TimvdLippe/servo/actions/runs/15083738556

TimvdLippe commented on tests/wpt/meta/preload/preload-csp.sub.html.ini:
```diff
@@ -0,0 +1,3 @@
+[preload-csp.sub.html]
```
It's strange to me that the `font-src` CSP tests pass, but this one doesn't. I double-checked if the request has all the correct information and added various `println!` statements in `rust-content-security-policy`, but wasn't able to figure out what we are still missing in `create_link_request`.

TimvdLippe replied:
It's because we don't support `track` requests yet. All the other requests are correctly blocked.

jdm commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -494,6 +507,128 @@ impl HTMLLinkElement {
+        // TODO
+        // Step 7. Let key be the result of creating a preload key given request.
+        // TODO
+        // Step 8. If options's document is "pending", then set request's initiator type to "early hint".
```
FYI: this step appears nonsensical to me. The spec may be attempting to refer to a session history entry, rather than a document: https://html.spec.whatwg.org/multipage/#she-step

jdm replied:
And as usual @simonwuelker is ahead of me: https://github.com/whatwg/html/issues/10640

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15095179190

This PR is now ready for review. Some of the WPT tests are not working, but I expect to work when we figure out the timing problem with `performance` entries.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#52624) with upstreamable changes.

TimvdLippe committed:
Move mime classifier to net_traits

That way, we can depend on it from the script package.

Also update several of the mime matching methods to match
the specification and add more relevant media types to it.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Implement basics of link preloading

These changes allow a minimal set of checks for font-src
CSP checks to pass.

Part of #4577
Part of #35035

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15095884918

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

simonwuelker commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -795,3 +1039,77 @@ impl PreInvoke for PrefetchContext {
 }
+
+struct PreloadContext {
+    /// The `<link>` element that caused this prefetch operation
+    link: Trusted<HTMLLinkElement>,
+
+    resource_timing: ResourceFetchTiming,
+
+    /// The url being prefetched
+    url: ServoUrl,
```
nit: please change "prefetch" to "preload" in these two comments, to avoid confusion with `<link rel=prefetch>`.

TimvdLippe commented on components/shared/net/tests/parsable_mime/text/xml/feed.rss:
```diff
@@ -1,151 +1,151 @@
-      <pubDate>Tue, 19 Oct 2004 11:08:56 -0400</pubDate>
-    </item>
-  </channel>
+<?xml version="1.0" encoding="windows-1252"?>
+<rss version="2.0">
+  <channel>
+    <title>FeedForAll Sample Feed</title>
+    <description>RSS is a fascinating technology. The uses for RSS are expanding daily. Take a closer look at how various industries are using the benefits of RSS in their businesses.</description>
+    <link>http://www.feedforall.com/industry-solutions.htm</link>
+    <category domain="www.dmoz.com">Computers/Software/Internet/Site Management/Content Management</category>
```
By moving these files, git detected that it was using CRLF line endings. Presumably these files were added prior to `.gitattributes` being configured. Hence, git now reconfigured them and I had no other option than to accept.

simonwuelker replied:
Thanks ^^

TimvdLippe commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -446,6 +546,13 @@ impl HTMLLinkElement {
             None => "",
         };
 
+        let matches_media_environment = element.matches_environment(mq_str);
+        self.previous_media_environment_matched
+            .set(matches_media_environment);
+        if !matches_media_environment {
+            return;
+        }
+
```
Ah in fact, this is unrelated. I fixed it in the attribute changed callback already, so this is indeed obsolete.

TimvdLippe commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -494,6 +599,128 @@ impl HTMLLinkElement {
+        if mime_type.is_empty() {
+            return true;
+        }
+        let Some(destination) = destination else {
+            return false;
+        };
```
It's because in the options `destination: Some<Destination>`. I changed that to be just a `Destination`, as we already have `Destination::None` to cover the empty case.

TimvdLippe commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -494,6 +599,128 @@ impl HTMLLinkElement {
+        let Some(destination) = destination else {
+            return false;
+        };
+        // Step 2. If destination is "fetch", then return true.
+        //
+        // Fetch is handled as an empty string destination in the spec
+        if destination == Destination::None {
+            return true;
+        }
```
No, this is not a bug in the spec. It's specified here: https://fetch.spec.whatwg.org/#concept-potential-destination-translate Will add that to the comment

TimvdLippe commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -494,6 +599,128 @@ impl HTMLLinkElement {
+            // Step 4. If mimeTypeRecord is failure, then return false.
+            return false;
+        };
+        // Step 5. If mimeTypeRecord is not supported by the user agent, then return false.
+        let Some(mime_type) = MimeClassifier::get_media_type(&mime_type_record) else {
+            return false;
+        };
```
Okay, I understand. Do you mean I need to update the code here or this is a clarifying comment?

simonwuelker replied:
I think the current behaviour is fine, but a clarifying comment would be nice

TimvdLippe replied:
Okay done 👍

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15145203412

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15145681714

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

TimvdLippe commented:
Reverted back the optional destination change, as that appears to break a lot of WPT tests. Confirming we are back green now: https://github.com/TimvdLippe/servo/actions/runs/15274995022

TimvdLippe commented:
Ah, according to https://disabled-wpt.goose.icu/#filter=preload%2Fpreload-error.sub.html it is indeed flaky :/

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52624).

TimvdLippe commented:
That WPT failure is a flake per my above comment. It also flakes in other browsers. Can we mark it as intermittent and file an issue with WPT to get it addressed?

simonwuelker commented:
> That WPT failure is a flake per my above comment. I

Sorry for missing this. I filed https://github.com/servo/servo/issues/37177.

simonwuelker commented:
@jdm is there a way to mark a test as having both intermittent PASSes and FAILs or do I need to open a second issue?

mrobinson commented:
> @jdm is there a way to mark a test as having both intermittent PASSes and FAILs or do I need to open a second issue?

You can add `expected: [PASS, FAIL]` to the `.ini` metadata for the test, which should allow it to accept both results.

jdm commented:
When an issue is not being recognized by our systems and merges are failing, usually reopening the issue or readding the label will help.
